trace-recorder/src/main/org/jetbrains/lincheck/trace/recorder/TraceCollectingEventTracker.kt

```diff
@@ -155,6 +157,14 @@ class TraceCollectingEventTracker(
     // only at the end
     private val threads = ConcurrentHashMap<Thread, ThreadData>()
 
+    // For proper completion of threads which are not tracked from the start of the agent,
+    // of of those threads which are not joined by the Main thread,
+    // we need to perform operations in them under locks.
+    // Note: the only place where there is a contention on the locks is
+    // when Main thread finishes and decides to "finish" all other running threads.
+    // By "finish" here we imply that it will dump their recorded data.
+    private val locks = ConcurrentHashMap<Thread, ReentrantLock>()
```

Comment:
This should be rewritten to thread-local variable + registering the link to the instance in some list. So that each read of the lock does not read from shared-memory and the list is required for synchronization when the Main Thread is completed and wants to disable tracking in all still-running traces

Reply:
We decided to implement a fine-grained flag-based solution instead